muxers/mplex/src/io.rs

```diff
@@ -0,0 +1,793 @@
+                return Poll::Pending
+            }
+        }
+
+        Poll::Ready(Ok(()))
+    }
+
+    /// Records a fatal error for the multiplexed I/O stream.
+    fn on_error<T>(&mut self, e: io::Error) -> io::Result<T> {
+        log::debug!("Multiplexed connection failed: {:?}", e);
+        self.status = Status::Err(io::Error::new(e.kind(), e.to_string()));
+        self.pending_frames =  Default::default();
+        self.open_substreams = Default::default();
+        self.buffer = Default::default();
+        Err(e)
+    }
+
+    /// Checks that the multiplexed stream has status `Ok`,
+    /// i.e. is not closed and did not encounter a fatal error.
+    fn guard_open(&self) -> io::Result<()> {
+        match &self.status {
+            Status::Closed => Err(io::Error::new(io::ErrorKind::Other, "Connection is closed")),
+            Status::Err(e) => Err(io::Error::new(e.kind(), e.to_string())),
+            Status::Open => Ok(())
+        }
+    }
+
+    /// Checks that the permissible limit for pending outgoing frames
+    /// has not been reached.
+    fn check_max_pending_frames(&mut self) -> io::Result<()> {
+        if self.pending_frames.len() >= self.config.max_substreams + EXTRA_PENDING_FRAMES {
+            return self.on_error(io::Error::new(io::ErrorKind::Other,
+                "Too many pending frames."));
+        }
+        Ok(())
+    }
+}
+
+/// The operating states of a substream.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+enum SubstreamState {
+    /// An `Open` frame has been received or sent.
+    Open,
+    /// A `Close` frame has been sent, but the stream is still open
+    /// for reading (half-close).
+    SendClosed,
+    /// A `Close` frame has been received but the stream is still
+    /// open for writing (remote half-close).
+    RecvClosed
+}
+
+struct NotifierRead {
+    /// List of wakers to wake when read operations can proceed
+    /// on a substream (or in general, for the key `None`).
+    pending: Mutex<FnvHashMap<Option<LocalStreamId>, Waker>>,
+}
+
+struct NotifierWrite {
+    /// List of wakers to wake when write operations on the
+    /// underlying I/O stream can proceed.
+    pending: Mutex<Vec<Waker>>,
+}
+
+impl NotifierWrite {
+    /// Registers interest of a task in writing to some substream.
+    ///
+    /// The returned waker should be passed to an I/O write operation
+    /// that schedules a wakeup, if necessary.
+    #[must_use]
+    fn register<'a>(self: &'a Arc<Self>, waker: &Waker) -> WakerRef<'a> {
+        let mut pending = self.pending.lock();
+        if pending.iter().all(|w| !w.will_wake(waker)) {
+            pending.push(waker.clone());
+        }
+        waker_ref(self)
+    }
+}
+
+impl ArcWake for NotifierWrite {
+    fn wake_by_ref(this: &Arc<Self>) {
+        let wakers = mem::replace(&mut *this.pending.lock(), Default::default());
+        for waker in wakers {
+            waker.wake();
+        }
+    }
+}
+
+impl NotifierRead {
```

Comment:
Any particular reason why the `NotifierRead` `impl`s are separated from the corresponding `struct ` definition?

Reply:
What do you mean by "separated"? That it is not directly below the `struct` definition? No reason for that, other than that I don't particularly mind about such ordering of definitions. I'm happy to change it.

Reply:
I am just a bit confused by the fact that all of `NotifierWrite` is interleaved. I am fine either way.

Reply:
I changed the ordering in https://github.com/libp2p/rust-libp2p/pull/1769/commits/77ea68b4f24ac77b0392d93a42c8cf4eb66d54ac.